### PR TITLE
Polish Pass: Utilities

### LINKED ISSUE #922: Polish Pass: Utility Features
> **Warning** - this ticket is a work in progress. If you have suggestions please add a comment below!

Base on:
- https://github.com/skeletonlabs/skeleton/issues/782

## Goals

The goal of this pass will be to improve, refine, and finalize all Utility features in Skeleton.

- [x] Fix and address known issues and bugs
- [x] Refactor code, cleanup styles, and improve edge cases
- [x] Improve documentation for each feature
- [x] Expand with new features as relevant
- [x] Make overlays not suck, or at least suck less
- [x] Ensure Vitest test cases are passing
- [x] Resolve `FIXME` issues ([reference](https://user-images.githubusercontent.com/1509726/216708976-ea8c37a1-adbf-46e5-a09f-095402907c6e.png))
- [x] Update all style props to use `CssClasses` type

### Local Storage Stores

- [x] Update to the latest upstream version 0.40.0

### Code Block

- [x] Minor visual refresh and code refactor
- [x] Property `buttonCopy` renamed `button`, default value is now `variant-soft`
- [x] Added `buttonLabel` and `buttonCopied` for basic i18n support
- [x] Added style props for `blur` and `shadow`
- [x] Prepare documentation "previewer" component

### Drawers

- [x] Docs: make it clear these are a singleton
- [x] Greatly improved documentation and setup instructions
- [x] Provide advanced tip for animating background contents, per [#637](https://github.com/skeletonlabs/skeleton/issues/637)
- [x] Refactor of Drawer animation logic
- [x] Added `shadow` property for Drawer element
- [x] Added `regionBackdrop|Drawer` props. Available as overrides as well!
- [x] BREAKING: Removed drawer `margin` in favor of backdrop `padding` (provides better results)
- [x] Rounded corners by default, based on theme settings (set `rounded-none` to disable)

### Lightswitch

> **Warning** - this utility feature has been complete rebuilt from the ground up.

- [x] Lightswitch component rebuilt with new UI design and new props available
- [x] Now uses a service-like approach, per [#737](https://github.com/skeletonlabs/skeleton/issues/737)
- [x] Provided an alternative to to the component if you wish to use OS preference automatically
- [x] Users can now interface with our methods, or build their own Lightswitch UI
- [x] Provided and documented interface stores, getters, and setter methods.

> NOTE: we need to test [#905](https://github.com/skeletonlabs/skeleton/issues/905)

### Modals

- [x] Docs: make it clear these are a singleton
- [x] Implement system to register and reuse custom modals, per [#720](https://github.com/skeletonlabs/skeleton/issues/720)
- [x] Custom modals should be "naked" by default and allow FULL customization
- [x] Allow prompt input to accept custom validation attributes, per [#899](https://github.com/skeletonlabs/skeleton/issues/899)
- [x] Added image modal example
- [x] Move modal example components to `/docs` - should not be part of the public package
- [x] Provide prop to set position - ex: top (search) or center (default), ([reference](https://user-images.githubusercontent.com/31982384/214623340-22252a12-fe33-42c8-9906-c40d385df4b8.png))
- [x] Improve component scrolling, per [#890](https://github.com/skeletonlabs/skeleton/issues/890)

> NOTE: given the potential scale and scope of changes required, we'll have to circle back to [#937](https://github.com/skeletonlabs/skeleton/issues/937)

### Toasts

- [x] Docs: make it clear these are a singleton
- [x] Wrapping button element needs selector like `.toast-actions`
- [x] Update button style props to allow changing button type
- [x] Added new `buttonDismissLabel` prop to override the dismiss button label text
- [x] Replaced `preset` prop with `background` property allowing any variant or background class.
- [x] Improved animation using Svelte FLIP 
- [x] Added a `callback` function feature for the ToastSettings object

### Popups

> **Warning** - this completely replaces Menus and Tooltip utilities
> Requires new dependency: https://floating-ui.com/

- [x] Replaced Menus and Tooltips with new "Popup" utility feature
- [x] Fully documented, has tons of config options
- [x] Updated examples, including a functional Combobox
- [x] Includes a fade animation on open/close
- [x] Supports multiple interaction event handlers (click, hover, both)
- [x] Can optionally include a customizable arrow on any variation
- [x] Now includes 12 position options
- [x] Dramatically better performance in keeping the popup on screen
- [x] Fully tree shakable
- [x] Allows you to modify a host of settings through "middleware"
- [x] Supports a11y features, such as keyboard interaction

### Menus

> DEPRECATED: replaced with Popups (shown above)

### Menus

> DEPRECATED: replaced with Popups (shown above)

### Data Tables

> See our our audit of prosed third party solutions here [#821](https://github.com/skeletonlabs/skeleton/issues/821)

### PR DESCRIPTION
**Pull request by endigo9740**
Fixes #922

### CODE REVIEW

File: src/lib/utilities/popup/types.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Popup Types
+// Note: these are a simple iteration based on the official docs.
+
+// Placement
+// https://floating-ui.com/docs/computePosition#placement
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:5]: Change to a `/**` comment to enable intellisense inclusion.

File: src/lib/utilities/popup/types.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	/** Arrow options: https://floating-ui.com/docs/arrow */
+	arrow?: Record<string, any>;
+}
+
+// Action Arguments
+export interface PopupSettings {
+	/** Provide the event type. */
+	event: 'click' | 'hover' | 'hover-click';
+	/** Match the popup data value `[data-popup]="targetNameHere"` */
+	target: string;
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:37]: If it's always going to be using the `data-popup` data attribute then we can type it as such, interpolating with a type

```ts
target: `[data-popup]="${string}"`

File: src/lib/utilities/popup/types.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Popup Types
+// Note: these are a simple iteration based on the official docs.
+
+// Placement
+// https://floating-ui.com/docs/computePosition#placement
+type Placement =
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:6]: through the magic of template literals, you can get this to be a little more concise.

```suggestion
type Direction = 'top' | 'bottom' | 'left' | 'right'
type Placement = Direction | `${Direction}-start` | `${Direction}-end`
```

File: src/lib/utilities/popup/types.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	/** Arrow options: https://floating-ui.com/docs/arrow */
+	arrow?: Record<string, any>;
+}
+
+// Action Arguments
+export interface PopupSettings {
+	/** Provide the event type. */
+	event: 'click' | 'hover' | 'hover-click';
+	/** Match the popup data value `[data-popup]="targetNameHere"` */
+	target: string;
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:37]: If it's always going to be using the `data-popup` data attribute then we can type it as such, interpolating with a type

```suggestion
    target: `[data-popup]="${string}"`
```

File: src/lib/utilities/popup/types.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+interface Middleware {
+	/** Offset options: https://floating-ui.com/docs/offset */
+	offset?: number | Record<string, any>;
+	/** Shift options: https://floating-ui.com/docs/shift */
+	shift?: Record<string, any>;
+	/** Flip options: https://floating-ui.com/docs/flip */
+	flip?: Record<string, any>;
+	/** Arrow options: https://floating-ui.com/docs/arrow */
+	arrow?: Record<string, any>;
+}
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:30]: (VSCode is bugging out, hope that this works, it's meant to be the one where it imports from floating UI)

```suggestion
interface Middleware {
	/** Offset options: https://floating-ui.com/docs/offset */
	offset?: OffsetOptions;
	/** Shift options: https://floating-ui.com/docs/shift */
	shift?: ShiftOptions;
	/** Flip options: https://floating-ui.com/docs/flip */
	flip?: FlipOptions;
	/** Arrow options: https://floating-ui.com/docs/arrow */
	arrow?: ArrowOptions;
}
```

File: src/lib/utilities/popup/types.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// Popup Types
+// Note: these are a simple iteration based on the official docs.
+
```
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:3]:
```suggestion
import type { ArrowOptions, FlipOptions, OffsetOptions, ShiftOptions } from '@floating-ui/dom';
```

File: src/lib/utilities/Modal/types.ts [hunk 7]
```diff
@@ -12,16 +12,20 @@ export interface ModalComponent {
...
-	image?: URL | string;
-	/** Prompt Only - provide a prompt value. */
-	value?: unknown;
-	/** Provide your component reference, props, and slot details. */
-	component?: ModalComponent;
+	image?: string;
+	/** By default, used to provide a prompt value. */
+	value?: any;
+	/** Provide input attributes as key/value pairs. */
+	valueAttr?: Object;
```
Comment (ryceg) [inline src/lib/utilities/Modal/types.ts:26]: From the eslint: 

> Don't use `Object` as a type. The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.
> - If you want a type meaning "any object", you probably want `object` instead.
> - If you want a type meaning "any value", you probably want `unknown` instead.

```suggestion
	valueAttr?: object;
```

**Review by ryceg — COMMENTED** [review]
I hope that you don't mind Chris, I've just gone through and made a couple suggestions for the typings- notably, you can actually pull in the types that are used by FloatingUI to put the monkey back on their back to make sure that the middleware documentation is up to standards. Feel free to ignore, some are just nit picking code quality like SemanticNames, which wasn't even part of the code changed.

File: src/lib/utilities/popup/types.ts (hunk 2 above)
Comment (endigo9740) [reply src/lib/utilities/popup/types.ts:37]: Perhaps the docs aren't clear here, but the use doesn't provide all that, they only put the actual value between the quotes for the data attribute.

So if I have:

`<div data-popup="foo">`

The target value will be:

`target: 'foo'`

Make sense? In this case I think string would make more sense.

File: src/lib/utilities/popup/types.ts (hunk 5 above)
Comment (ryceg) [inline src/lib/utilities/popup/types.ts:30]:
```suggestion
interface Middleware {
	/** Offset options: https://floating-ui.com/docs/offset */
	offset?: number | Record<string, any>;
	/** Shift options: https://floating-ui.com/docs/shift */
	shift?: Record<string, any>;
	/** Flip options: https://floating-ui.com/docs/flip */
	flip?: Record<string, any>;
	/** Arrow options: https://floating-ui.com/docs/arrow */
	arrow?: {
		element: string;
	} & Record<string, any>;
}
```

Comment (endigo9740) [reply src/lib/utilities/popup/types.ts:30]: Good catch! Copy and paste strikes again!

File: src/lib/utilities/Toast/types.ts [hunk 8]
```diff
@@ -8,6 +8,8 @@ export interface ToastSettings {
 	 * @type {'primary' | 'secondary' | 'tertiary' | 'warning' | 'success' | 'error'}
 	 */
 	preset?: 'primary' | 'secondary' | 'tertiary' | 'warning' | 'success' | 'error';
```
[the remark this answers is not in the document]
Comment (endigo9740) [reply src/lib/utilities/Toast/types.ts:10]: The `present` option has been replaced with `background`, I just forgot to nuke it here. Thanks for the reminder!

File: src/lib/utilities/popup/types.ts (hunk 4 above)
Comment (endigo9740) [reply src/lib/utilities/popup/types.ts:37]: See the other message regarding this

File: src/lib/utilities/popup/types.ts (hunk 6 above)
Comment (endigo9740) [reply src/lib/utilities/popup/types.ts:3]: Unfortunately as discussed on Discord these are not possible to use within the library files.

**Comment by endigo9740** [comment]
@ryceg I've applied all above feedback, aside from one using the imported types from Floating UI.

File: src/lib/utilities/Toast/types.ts [hunk 9]
```diff
@@ -21,6 +21,8 @@ export interface ToastSettings {
 	};
 	/** Provide arbitrary CSS classes to style the toast. */
 	classes?: string;
+	/** Callback function that fires on trigger and close. */
+	callback?: (response: { id: string; status: string }) => void;
```
Comment (AdrianGonz97) [inline src/lib/utilities/Toast/types.ts:25]: Small change since the status can only be either "closed" or "queued" so we can make it a union type.
```suggestion
	callback?: (response: { id: string; status: 'queued' | 'closed' }) => void;
```

**Comment by endigo9740** [comment]
FYI this has been merged into `dev` branch, please continue QA testing there!
https://dev.skeleton.dev/
